### ISSUE TITLE
Error when using nil in hash values with includes method

### ISSUE THREAD

**Issue by akicho8** [rails/rails#52592]
### Steps to reproduce

I have been using a convention where I specify all arguments for the includes method in hash format to avoid complexity.
For example, I would write `includes(a: nil, b: :c)` instead of `includes(:a, b: :c)`.
This approach has worked well for me for years without issues.

However, I encountered an error when calling the `ids` method if the hash value contains `nil`.

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  # gem "rails"
  # If you want to test against edge Rails replace the previous line with this:
  gem "rails", github: "rails/rails", branch: "main"

  gem "sqlite3", "~> 1.4"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)
ActiveSupport::LogSubscriber.colorize_logging = false

ActiveRecord::Schema.define do
  create_table :posts do |t|
  end
  create_table :comments do |t|
    t.belongs_to :post
  end
end

class Post < ActiveRecord::Base
  has_many :comments
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!

    assert_equal [1], Post.includes(:comments).to_a.collect(&:id)
    assert_equal [1], Post.includes(comments: nil).to_a.collect(&:id)

    assert_equal [1], Post.includes(:comments).ids
    assert_equal [1], Post.includes(comments: nil).ids
  end
end
# >> Fetching https://github.com/rails/rails.git
# >> Fetching gem metadata from https://rubygems.org/..........
# >> Resolving dependencies...
# >> -- create_table(:posts)
# >> D, [2024-08-13T21:27:56.400153 #2203] DEBUG -- :    (1.9ms)  CREATE TABLE "posts" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL)
# >>    -> 0.0043s
# >> -- create_table(:comments)
# >> D, [2024-08-13T21:27:56.400420 #2203] DEBUG -- :    (0.0ms)  CREATE TABLE "comments" ("id" integer PRIMARY KEY AUTOINCREMENT NOT NULL, "post_id" integer)
# >> D, [2024-08-13T21:27:56.400564 #2203] DEBUG -- :    (0.0ms)  CREATE INDEX "index_comments_on_post_id" ON "comments" ("post_id")
# >>    -> 0.0003s
# >> D, [2024-08-13T21:27:56.401155 #2203] DEBUG -- :    (0.0ms)  CREATE TABLE "schema_migrations" ("version" varchar NOT NULL PRIMARY KEY)
# >> D, [2024-08-13T21:27:56.401852 #2203] DEBUG -- :    (0.1ms)  CREATE TABLE "ar_internal_metadata" ("key" varchar NOT NULL PRIMARY KEY, "value" varchar, "created_at" datetime(6) NOT NULL, "updated_at" datetime(6) NOT NULL)
# >> D, [2024-08-13T21:27:56.414546 #2203] DEBUG -- :   ActiveRecord::InternalMetadata Load (1.2ms)  SELECT * FROM "ar_internal_metadata" WHERE "ar_internal_metadata"."key" = ? ORDER BY "ar_internal_metadata"."key" ASC LIMIT 1  [[nil, "environment"]]
# >> D, [2024-08-13T21:27:56.414763 #2203] DEBUG -- :   ActiveRecord::InternalMetadata Create (0.1ms)  INSERT INTO "ar_internal_metadata" ("key", "value", "created_at", "updated_at") VALUES ('environment', 'development', '2024-08-13 12:27:56.414585', '2024-08-13 12:27:56.414588') RETURNING "key"
# >> Run options: --seed 57747
# >> 
# >> # Running:
# >> 
# >> D, [2024-08-13T21:27:56.454822 #2203] DEBUG -- :   SQL (0.1ms)  BEGIN immediate TRANSACTION
# >> D, [2024-08-13T21:27:56.454913 #2203] DEBUG -- :   Post Create (0.2ms)  INSERT INTO "posts" DEFAULT VALUES RETURNING "id"
# >> D, [2024-08-13T21:27:56.455014 #2203] DEBUG -- :   TRANSACTION (0.0ms)  COMMIT TRANSACTION
# >> D, [2024-08-13T21:27:56.459111 #2203] DEBUG -- :   Post Load (0.0ms)  SELECT "posts".* FROM "posts"
# >> D, [2024-08-13T21:27:56.473230 #2203] DEBUG -- :   Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = ?  [["post_id", 1]]
# >> D, [2024-08-13T21:27:56.474008 #2203] DEBUG -- :   Post Load (0.0ms)  SELECT "posts".* FROM "posts"
# >> D, [2024-08-13T21:27:56.474204 #2203] DEBUG -- :   Comment Load (0.0ms)  SELECT "comments".* FROM "comments" WHERE "comments"."post_id" = ?  [["post_id", 1]]
# >> D, [2024-08-13T21:27:56.479940 #2203] DEBUG -- :   Post Ids (0.5ms)  SELECT "posts"."id" FROM "posts" LEFT OUTER JOIN "comments" ON "comments"."post_id" = "posts"."id" GROUP BY "posts"."id"
# >> E
# >> 
# >> Finished in 0.033922s, 29.4794 runs/s, 88.4382 assertions/s.
# >> 
# >>   1) Error:
# >> BugTest#test_association_stuff:
# >> ActiveRecord::ConfigurationError: nil
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:67:in `walk_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:64:in `block in walk_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:62:in `each'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:62:in `walk_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:59:in `block in walk_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:58:in `each'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:58:in `walk_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:49:in `make_tree'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/associations/join_dependency.rb:72:in `initialize'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/relation/query_methods.rb:1597:in `new'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/relation/query_methods.rb:1597:in `construct_join_dependency'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/relation/finder_methods.rb:458:in `apply_join_dependency'
# >>     /opt/rbenv/versions/3.2.2/lib/ruby/gems/3.2.0/bundler/gems/rails-a32822f6ae15/activerecord/lib/active_record/relation/calculations.rb:382:in `ids'
# >>     -:48:in `test_association_stuff'
# >> 
# >> 1 runs, 3 assertions, 0 failures, 1 errors, 0 skips
```

### Expected behavior

```ruby
assert_equal [1], Post.includes(comments: nil).ids
```

### Actual behavior

```ruby
assert_equal [1], Post.includes(comments: nil).ids # => #<ActiveRecord::ConfigurationError: nil>
```

### System configuration

```ruby
ActiveRecord::VERSION::STRING   # => "7.1.3.4"
```

**Comment by fatkodima** [comment]
Can you provide a reproduction test script using this template - https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_record.rb?

What was the latest rails version when it worked?

**Comment by akicho8** [comment]
@fatkodima Thank you for viewing the issue. I rewrote it using a template.

I'm not sure if it worked in previous versions or if it has always been this way.
That's because, usually, I call `ids` on relations without executing `includes`.
I discovered this behavior by accident this time.

**Comment by jianbo** [comment]
Maybe this after this https://github.com/rails/rails/pull/46697

**Comment by fatkodima** [comment]
I just tested with rails versions >= 6.1, and it does not work. So, are you reporting a bug (something worked before and now it is not working; if yes, on which rails version it worked) or asking for a feature request?

**Comment by akicho8** [comment]
I am switching back and forth between English and Japanese in translation, so I apologize if there are any discrepancies.

First, I am not sure if `includes(comments: nil).ids` used to work.
I don't know if `includes(includes: nil).ids` worked before, partly because `includes` and `ids` are conflicting functions, and I rarely used a combination of the two.
I only noticed it this time because I was debugging and trying to check the status of a complex relation.

In other words, I am not saying that the application stopped working when I upgraded to a newer version of Rails, but rather that I am noticing this behavior for the first time.

Therefore, I am not exactly sure if this is a "bug report" or a "request".

However, personally, it was quite difficult to determine the cause of this error, so regardless of whether it is a "bug report" or a "request", I would like to see `Post.includes(comments: nil).to_a.collect(&:id)` and `Post.includes(comments: nil).ids` have the same result.
